wrappers/llvm-mingw_toolchainfile.cmake.in

```diff
@@ -0,0 +1,84 @@
+if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.20")
+  set(_prefix "${CMAKE_CURRENT_LIST_DIR}/@RELPATH@/")
+  cmake_path(ABSOLUTE_PATH _prefix NORMALIZE)
+else()
+  get_filename_component(_prefix "${CMAKE_CURRENT_LIST_DIR}/@RELPATH@/" ABSOLUTE)
+endif()
+
+if(NOT CMAKE_SYSTEM_NAME)
+  set(CMAKE_SYSTEM_NAME Windows)
```

Comment:
Thats pretty vague, cross-compiling is not just a different System.

-   If I run x64 Windows and compile for Arm Windows I do a cross-compilation.
-   What if you dont have UCRT installed, but the toolchain uses UCRT.
-   On Linux you have tons of more differences - different flavors of Linux and runtimes.

I get that you potentially could safe some time, but its not trivial to guarantee the stuff will run on the host and the user will then have a broken build with potentially no useful error message on Windows (like the needed DLLs are not in PATH).

Way too many pitfalls and downsides for me.

The most reliably way to build for host would be to run CMake without params and hope the User prepared for that. (Or use a package manager like Conan which will also cache the tools once they are built).

Reply:
I think that it would be important to not set cmake into cross compiling mode when not necessary. I agree that it's not trivial though. One problem is that `CMAKE_HOST_SYSTEM_PROCESSOR` mostly indicates what architecture CMake itself was compiled as.

I believe the logic could be something like this (in pseudocode):
```
if (CMAKE_HOST_SYSTEM_NAME == "Windows")
    if (CMAKE_HOST_SYSTEM_PROCESSOR == ${arch})
        # Exact architecture match, not cross compiling
    elif (CMAKE_HOST_SYSTEM_PROCESSOR == x86_64 AND ${arch} == i686)
        # Differing architectures, but should be executable
    elif (CMAKE_HOST_SYSTEM_PROCESSOR == aarch64 AND ${arch} == armv7)
        # Differing architectures, but should be executable. (Although future aarch64 processors may remove 32 bit support.)
    else
        # Cannot execute the built executables, indicate that we're cross compiling.
        set(CMAKE_SYSTEM_NAME "Windows")
    endif
else
    set(CMAKE_SYSTEM_NAME "Windows")
endif()
```

Yes, this is not water tight for some potential odd combinations of CRT and such, but I think this would be a better user experience (especially wrt compiling something like LLVM itself) rather than always flagging it as cross compilation.

Reply:
> I think that it would be important to not set cmake into cross compiling mode when not necessary.

I think you overestimate the amount of Projects depending on `CMAKE_CROSSCOMPILING`, the meaning/usage of the variable seems not fully clear even among CMake Maintainers.
I believe this is legacy Stuff from decades ago.

Its recommended to write tests for compiler features, and you can do the same for stuff you plan to run:

``` CMake
include(CheckCSourceRuns)
check_c_source_runs("int main() {return 0;}" HEY_I_CAN_RUN_THE_STUFF_I_BUILD)
```

This gives you precisely the info you need (but this only works after the `project` command and after the toolchain is consumed).

So that's a Project problem first and foremost and the cost is some inefficiency, the adverse effects of `CMAKE_CROSSCOMPILING` missing when it should be set is a broken build (potentially after hours).
Even then I think it would be more sense to set `CMAKE_CROSSCOMPILING`  and use `CMAKE_CROSSCOMPILING_EMULATOR` to setup `PATH`.

> I believe the logic could be something like this (in pseudocode):

I'm strictly against adding such guesswork. if you really want then I add a *separate* toolchain for "native" builds.

I had planned to add logic to optionally use a system installed clang in-place, but I'd want simple reliable toolchains as base.
ie. you could extend the toolchain with our logic:

``` CMake
include("${CMAKE_CURRENT_LIST_DIR}/llvm-mingw_toolchainfile.cmake")
if-but-else-if-I-think-it-could-run-native()
  unset(CMAKE_SYSTEM_NAME)
endif()
```

Reply:
> I think you overestimate the amount of Projects depending on `CMAKE_CROSSCOMPILING`, the meaning/usage of the variable seems not fully clear even among CMake Maintainers.
> I believe this is legacy Stuff from decades ago.

I disagree with your belief. Yes it's not a huge amount, but as someone who has spent a significant amount of time in the CMake files of LLVM and its runtimes, I believe getting this right is kinda vital.

> Its recommended to write tests for compiler features, and you can do the same for stuff you plan to run:
> 
> ```cmake
> include(CheckCSourceRuns)
> check_c_source_runs("int main() {return 0;}" HEY_I_CAN_RUN_THE_STUFF_I_BUILD)
> ```
> 
> This gives you precisely the info you need (but this only works after the `project` command and after the toolchain is consumed).

Feel free to use such constructs in your CMake projects, but in the ones I've touched, `CMAKE_CROSSCOMPILING` is the key.

> I'm strictly against adding such guesswork. if you really want then I add a _separate_ toolchain for "native" builds.

No, a separate toolchain file sounds much worse. You may dislike the guesswork, but that is the form that it'll be for it to be merged here.

Reply:
Another aspect which admittedly makes the comparison between our arch name and `CMAKE_HOST_SYSTEM_PROCESSOR` a bit more messy, is that the latter doesn't use names in the same namespace, so they'll have to be remapped. In general, `CMAKE_HOST_SYSTEM_PROCESSOR` across OSes is not consistent either, as it on Unix falls back on `uname -m` while it uses the `PROCESSOR_ARCHITECTURE` or `PROCESSOR_ARCHITEW6432` env variables on Windows. But for this purpose, it's enough to check the names used on windows - `x86`, `AMD64`, `ARM64` (and presumably `ARM` for 32 bit there).

Reply:
> Yes it's not a huge amount of projects using it, but as someone who has spent a significant amount of time in the CMake files of LLVM and its runtimes, I believe getting this right is kinda vital.

Its unlikely that using a toolchain-file is a good fit for building
a multi-stage, potentially bootstrapping toolchain.

The amount of projects which would benefit from a toolchain-file,
require `CMAKE_CROSSCOMPILING` and need some magic detection in the
toolchain-file must be close to zero.

LLVM is not exactly an argument for toolchain usability here.

> I believe getting this right is kinda vital.

Yep, for me that means being useful for the vast majority of cases,
and as little surprises as possible.
I still don't get what you wanna do, use a toolchain-file with LLVM?
This sounds like a hypothetical idea. You dont *have* to use
them just because they are there.

> No, a separate toolchain file sounds much worse

There are already 9 toolchain-files?
One as basis, 4 per arch and 4 more for uwp (those 4 I would actually prefer to remove).

What I would propose:

1.  Dont set `CMAKE_SYSTEM_NAME` on windows in the base `llvm-mingw_toolchainfile.cmake`

    small change:

    ``` cmake 
    if(NOT CMAKE_SYSTEM_NAME AND NOT CMAKE_HOST_WIN32)
      set(CMAKE_SYSTEM_NAME Windows)
    endif()
    ```

2.  explicitly set `CMAKE_SYSTEM_NAME` in the others (like `x86_64-w64-mingw32_toolchainfile.cmake`)

    small addition:

    ``` cmake 
    set(CMAKE_SYSTEM_NAME Windows)
    ```

Then, if you want to build "native", use `llvm-mingw_toolchainfile.cmake`,
otherwise pick your arch.

If you have a complicated project where a toolchain-file helps,
you want a native build and `llvm-mingw_toolchainfile.cmake` does the wrong thing - set
`-DCMAKE_SYSTEM_NAME=Windows` manually.

Reply:
> Its unlikely that using a toolchain-file is a good fit for building a multi-stage, potentially bootstrapping toolchain.

Didn't you earlier argue that toolchain files are important especially for more complicated build setups like that?

> > I believe getting this right is kinda vital.
> 
> Yep, for me that means being useful for the vast majority of cases, and as little surprises as possible.

Yes, and having the cross compilation flag set, when we're in fact not cross compiling, sounds like an unnecessary surprise to me.

The code I'm suggesting is no 70% chance maybe-works heuristic - I don't foresee any surprises of setting it wrong.

If we can just set it correctly, without needing to bother the user with the choice at all, why not just do that? Just because it would amount to 10 lines of cmake code that you dislike?

> > No, a separate toolchain file sounds much worse
> 
> There are already 9 toolchain-files? One as basis, 4 per arch and 4 more for uwp (those 4 I would actually prefer to remove).

Yes, and adding 4 more to distinguish between cross compilation and native compilation sounds quite unnecessary.

> Then, if you want to build "native", use `llvm-mingw_toolchainfile.cmake`, otherwise pick your arch.
> 
> If you have a complicated project where a toolchain-file helps, you want a native build and `llvm-mingw_toolchainfile.cmake` does the wrong thing - set `-DCMAKE_SYSTEM_NAME=Windows` manually.

That sounds like absolutely terrible usability. Where are we supposed to document that, and would anybody read that document?

Reply:
> Didn't you earlier argue that toolchain files are important especially for more complicated build setups like that?

Yes, they can and *generally* do make things a-ton easier. Didn't you say you don't care for them?
Please, if you did not, then give the Testsuite project a spin with VScode - there is barely any setup while allowing countless integrations and that's the point. You can get there probably by adding countless CMake flags but not when wrappers hide the information.
Look at toolchain-files as re-usable documentation.

> Yes, and having the cross compilation flag set, when we're in fact not cross compiling, sounds like an unnecessary surprise to me.

Sure

> The code I'm suggesting is no 70% chance maybe-works heuristic - I don't foresee any surprises of setting it wrong.

Building a 32bit DLL and expecting it to work with 64bit Host programs? Not having the most basic DLLs in path?
This is my problem, not un-setting the cross compilation flag.

> If we can just set it correctly, without needing to bother the user with the choice at all, why not just do that? Just because it would amount to 10 lines of cmake code that you dislike?

The definition of *correctly* is the problem. How do you get the DLLs in the PATH, with tons of different IDEs?

> Yes, and adding 4 more to distinguish between cross compilation and native compilation sounds quite unnecessary.

I am not talking of 4 more, 5 in total. (If its ok with you i'll scrap the 4 UWP ones)

> > Then, if you want to build "native", use `llvm-mingw_toolchainfile.cmake`, otherwise pick your arch.
> > If you have a complicated project where a toolchain-file helps, you want a native build and `llvm-mingw_toolchainfile.cmake` does the wrong thing - set `-DCMAKE_SYSTEM_NAME=Windows` manually.
> 
> That sounds like absolutely terrible usability. Where are we supposed to document that, and would anybody read that document?

1.  If you want to build for your System, use `llvm-mingw_toolchainfile.cmake` (never `CMAKE_CROSSCOMPILING`)
2.  Otherwise, pick the one for the System you like (always `CMAKE_CROSSCOMPILING`)

And that should cover nearly every CMake Project, takes out the heuristic and whatever corner-cases are lurking.
If there is a need to revisit this, then figure out improvements.

Reply:
> > The code I'm suggesting is no 70% chance maybe-works heuristic - I don't foresee any surprises of setting it wrong.
> 
> Building a 32bit DLL and expecting it to work with 64bit Host programs? Not having the most basic DLLs in path? This is my problem, not un-setting the cross compilation flag.

Fair enough, having the right DLLs in path for 32 bit builds on 64 bit can be an issue.

> > Yes, and adding 4 more to distinguish between cross compilation and native compilation sounds quite unnecessary.
> 
> I am not talking of 4 more, 5 in total. (If its ok with you i'll scrap the 4 UWP ones)

Ok, I apparently misunderstood what you suggested at that point before.

> 1. If you want to build for your System, use `llvm-mingw_toolchainfile.cmake` (never `CMAKE_CROSSCOMPILING`)
> 2. Otherwise, pick the one for the System you like (always `CMAKE_CROSSCOMPILING`)

Ok, I guess that kind of could work. That's kind of analoguous to using the triple prefixed tools, `<arch>-w64-mingw32-clang` vs invoking just `clang` which should build something with the same arch as the toolchain itself.

One potential pitfall though, regarding the DLLs: If I run a i686 toolchain (where the LLVM binaries are i686, and likewise libc++.dll and others) on an x86_64 machine, the default architecture taken from `CMAKE_HOST_SYSTEM_PROCESSOR` will be x86_64. However the DLLs that are in path are for i686. (This is probably an unusual situation for normal users, but I often test such setups.)

So instead of deriving the default from `CMAKE_HOST_SYSTEM_PROCESSOR`, I think it'd be safer to derive it from `clang -dumpmachine` - what do you think?

Reply:
Well, there are ways around that, like picking the i686 toolchain-file.
But I guess you and I will be able to deal with the weirder stuff, and I would keep the "supported" and documented uses simple. Then see if you get reports.

To me native("system") would be 64bit here, users might not even know if the toolchain is compiled for 32bit.
There are 64bit ARM CPUs that dont run 32bit Code, so cross compiling and doing the "right thing" is subjective.

I would argue that the real issue is locating the fitting DLL (Paths), some general problem (#409). Thats my original point to err on caution and always use a separate host toolchain *by default* (even better a package manager).

if we would adopt your change I would rather patch the static cpu architecture into the toolchain-file during build.
We could also go halfway and just print a warning if running and compiled host cpu is different.

Reply:
> and I would keep the "supported" and documented uses simple.
> 
> To me native("system") would be 64bit here, users might not even know if the toolchain is compiled for 32bit.

Sorry, I disagree. I don’t find what you’re suggesting to be any simpler. Why do the wrong thing when you can do the right thing?

Each Windows toolchain in llvm-mingw has got a default architecture - which is what you get when you execute the unprefixed tools. This has the added property that you’ll find the right DLLs straight in `<toolchain>/bin`.

Here, I would expect that to be what to be used, when using the “default native” toolchain file, to be using the toolchain default arch, not dependent on what system happens to be reporting.

> There are 64bit ARM CPUs that dont run 32bit Code, so cross compiling and doing the "right thing" is subjective.

You’re missing the point here. If you’ve downloaded a 32 bit toolchain and run it on a 64 bit arm system, the program isn’t that you can’t execute the binaries you’ve just built. You won’t be able to run the compiler in the first place!

The toolchain default arch will always be a reasonable thing use, otherwise you’ve got a fatally incompatible toolchain build.

> if we would adopt your change I would rather patch the static cpu architecture into the toolchain-file during build.

Sorry, but that sounds unnecessarily inflexible. The contents of the supposedly arch independent files under `share` would suddenly have such details hardcoded.

> We could also go halfway and just print a warning if running and compiled host cpu is different.

Just no. Why go to all that length to check things, just to warn, rather than just pick up the architecture and use that, and silently do the right thing?

Reply:
> Here, I would expect that to be what to be used, when using the “default native” toolchain file, to be using the toolchain default arch, not dependent on what system happens to be reporting.

The issue here is that the toolchain is used in context of **CMake**, and to me whether clang is 32 or 64 bit is an implementation detail.
You now have 5 toolchain-files, 4 of them will give you the same result no matter which version you installed.

The other should have the same understanding as CMake - and commands file `find_package`, `find_library` for the HOST will be aligned to `CMAKE_HOST_SYSTEM_PROCESSOR`.
And by that targeting that, will also give you the same result no matter which clang version you installed.

> You’re missing the point here. If you’ve downloaded a 32 bit toolchain and run it on a 64 bit arm system, the program isn’t that you can’t execute the binaries you’ve just built. You won’t be able to run the compiler in the first place!

Its about defining what a "native" build means. I would argue that targeting a 32bit architecture from 64bit cross-compiling and no fit for the toolchain-file which we claimed to be "native".

> The toolchain default arch will always be a reasonable thing use, otherwise you’ve got a fatally incompatible toolchain build.

Unless you use CMake functions which will resolve to 64bit libraries for example.

> > if we would adopt your change I would rather patch the static cpu architecture into the toolchain-file during build.
> 
> Sorry, but that sounds unnecessarily inflexible. The contents of the supposedly arch independent files under `share` would suddenly have such details hardcoded.

Not sure we talk about the same thing here, I am talking about patching in the final value during running `install-wrappers.sh`. Not sure you would then later swap clang without the toolchain-files?

> Just no. Why go to all that length to check things, just to warn, rather than just pick up the architecture and use that, and silently do the right thing?

I dont get your motivition, you seem to not be inclined to use the toolchain-files, but try to cook up some rather weird setups where they fail and the result should be to add further complications.
I can cook up examples where your "right thing" will cause troubles, especially if the use the right thing within CMake to lookup DLLs.

The real question is: "Why would a user install a 32bit toolchain on a 64bit OS?"
Are there valid reason for this, because if not then this is just a unnecessary problematic mess tying to the definition of "cross-compiling".

I would add a check to test if thats the case, add a warning that "Your toolchain is not matching CMake's host architecture".
With the fineprint that the user should either be able to fix the almost certain troubles, or install the fitting toolchain.

Reply:
> The other should have the same understanding as CMake - and commands file `find_package`, `find_library` for the HOST will be aligned to `CMAKE_HOST_SYSTEM_PROCESSOR`. And by that targeting that, will also give you the same result no matter which clang version you installed.

No, `find_package` and `find_library` will be aligned with the settings for `CMAKE_FIND_ROOT_...`. For the strict cross compilation cases, this really shouldn't be looking into anything related to the host system. So why would it in this case, when the only difference is that we're not setting `CMAKE_SYSTEM_NAME`?

> > Sorry, but that sounds unnecessarily inflexible. The contents of the supposedly arch independent files under `share` would suddenly have such details hardcoded.
> 
> Not sure we talk about the same thing here, I am talking about patching in the final value during running `install-wrappers.sh`. Not sure you would then later swap clang without the toolchain-files?

When cross-building the toolchains, we have a couple of cases where we transplant pieces from one toolchain build into another. E.g. when we build the toolchains that will run on Windows, they are cross-compiled on Linux. We don't rebuild the mingw-w64 files with this new cross compiled toolchain (as it can't be executed on that host), but we copy over those compiled files the preexisting Linux based toolchain. So we take a bunch of directories from an existing toolchain and transplant them into a new one.

In that case, we'd still run `install-wrappers.sh` for the new toolchain and not copy over the output of that, so having the default arch hardcoded in the files wouldn't break any of the current build processes directly. But my point is that it is surprising and unnecessary to have the contents of `share` be build dependent in that way.

> I dont get your motivition, you seem to not be inclined to use the toolchain-files, but try to cook up some rather weird setups where they fail and the result should be to add further complications.

Sorry, but this sounds quite disrespectful.

If I am to merge and maintain toolchain files, I want them to behave in a way that I find reasonable and consistent. The definition of what that means, is up to me.

I don't want to add further complications needlessly - but if I feel that it would be important for what I consider reasonable behaviour, I wouldn't shy away from a little extra complication in one file.

In this case, what I'm suggesting wouldn't add any complication at all, it would remove the whole mapping of `CMAKE_HOST_SYSTEM_PROCESSOR` to our architecture names, replaced with taking the output of `clang -dumpmachine` and extracting the arch name from there. That's a simplification in my opinion.

You said yourself that you consider running a 32 bit toolchain on a 64 bit host a contrieved case that you don't care about. Then why do you even bother? Why do you want to add further complications to add a warning, when you can just make it use the arch name from that output and _not_ add a warning?

Reply:
> No, `find_package` and `find_library` will be aligned with the settings for `CMAKE_FIND_ROOT_...`. For the strict cross compilation cases, this really shouldn't be looking into anything related to the host system. So why would it in this case, when the only difference is that we're not setting `CMAKE_SYSTEM_NAME`?

The issue is not cross-compiling, but this one being the "native" toolchain-file.
The project might do a variety of things here (including spawning a process and calling CMake without any flags since well... its "native"), things are already tricky and I doubt changing the architecture helps.
TBH I dont develop an Windows, I dont know how lookup is working there but there are for ex. flags like `REGISTRY_VIEW HOST` which seem like an obvious violation.

> But my point is that it is surprising and unnecessary to have the contents of share be build dependent in that way.

I half agree here, I dont want the contents in share to be build dependent. But I would go further to the `function` not being build dependent, which it is if you fetch variables from clang.
The value is fixed for a finalized toolchain and in that case, I would prefer doing it once and explicitly having the result readable in plain text.

> Sorry, but this sounds quite disrespectful.

My apologies, I thought the remainder should clear that up. You bring up hypothetical problems and questions wheres I stated more than once that I would wait for explicit problems and start with a simple, least surprises toolchain.

Having a "native" toolchain-file that uses at best an ambiguous definition of that (Id call targeting 32bit on 64bit cross-compiling), while correctly targeted cross-compile toolchain-file's are available doesnt make sense to me.

> If I am to merge and maintain toolchain files, I want them to behave in a way that I find reasonable and consistent. The definition of what that means, is up to me.

I try to explain why that's the opposite, given that I explicitly use CMake and Toolchains for a while, including llvm-mingw and the proposed toolchains.

> In this case, what I'm suggesting wouldn't add any complication at all, it would remove the whole mapping of CMAKE_HOST_SYSTEM_PROCESSOR to our architecture names, replaced with taking the output of clang -dumpmachine and extracting the arch name from there. That's a simplification in my opinion.

It's **another** dynamic behavior on top of CMake's logic, potentially breaking assumptions.

> You said yourself that you consider running a 32 bit toolchain on a 64 bit host a contrieved case that you don't care about. Then why do you even bother? Why do you want to add further complications to add a warning, when you can just make it use the arch name from that output and not add a warning?

You brought the case up, and I'd rather want to infer feedback from users when they run into problems and what they want to try to do.
Cause I think that's the best think if you want to maintain the toolchain. I doubt its a problem, I doubt your solution will help and I am not keen on adding stuff I haven't seen done anywhere just on a hunch.

You heave the last word, but this is my recommendation for maintaining.

Reply:
Fetching dumpmachine would look like this:

``` cmake
if(NOT CMAKE_SYSTEM_PROCESSOR)
  set(CMAKE_SYSTEM_PROCESSOR ${CMAKE_HOST_SYSTEM_PROCESSOR})
  if(CMAKE_HOST_WIN32)
    execute_process(COMMAND "${_prefix}bin/clang.exe" -dumpmachine
      RESULT_VARIABLE cresult
      OUTPUT_VARIABLE _respath
      OUTPUT_STRIP_TRAILING_WHITESPACE
    )
    if(NOT cresult EQUAL 0)
      message(FATAL_ERROR "\"${_prefix}bin/clang.exe\" -dumpmachine: failed with ${cresult}")
    endif()
    unset(cresult)
    string(REGEX REPLACE "-.*" "" CMAKE_SYSTEM_PROCESSOR "${_respath}")
```

Do you know what `clang -dumpmachine` returns on arm? Is it `armv7` or something different?

Reply:
Thanks, that looks reasonable.

Yes, it returns `armv7` for those targets, and `i686` in for such toolchains (it's not normalized into `i386`) - so we should be able to reuse the literal spelling of the arch here without any remapping. (In this case, it's also possible to test this by running e.g. `armv7-w64-mingw32-clang -dumpmachine` and see what that outputs.)